components/constellation/constellation.rs

```diff
@@ -2050,33 +2119,39 @@ impl<Message, LTF, STF> Constellation<Message, LTF, STF>
             }
         }
 
-        if self.frames.contains_key(&frame_change.frame_id) {
-            if frame_change.replace {
-                let evicted = self.frames.get_mut(&frame_change.frame_id).map(|frame| {
-                    frame.replace_current(frame_change.new_pipeline_id)
-                });
-                if let Some(evicted) = evicted {
-                    self.close_pipeline(evicted.pipeline_id, ExitPipelineMode::Normal);
-                }
-            } else {
-                if let Some(ref mut frame) = self.frames.get_mut(&frame_change.frame_id) {
-                    frame.load(frame_change.new_pipeline_id);
-                }
-            }
+        let (evicted_id, new_frame, clear_future, location_changed) = if let Some(mut entry) = frame_change.replace {
+            debug!("Replacing pipeline in existing frame.");
+            let evicted_id = entry.pipeline_id;
+            entry.pipeline_id = Some(frame_change.new_pipeline_id);
+            self.traverse_to_entry(entry);
+            (evicted_id, false, false, false)
+        } else if let Some(frame) = self.frames.get_mut(&frame_change.frame_id) {
+            debug!("Adding pipeline to existing frame.");
+            frame.load(frame_change.new_pipeline_id, frame_change.url.clone());
+            let evicted_id = frame.prev.get_mut(opts::get().max_session_history)
```

Comment:
At the moment I'm just handling discarding the past, since the future is discarded whenever navigation happens, so you really have to go out of your way to build up significant session future.